app/src/main/java/agency/five/codebase/android/movieapp/data/MovieRepositoryImpl.kt

```diff
@@ -0,0 +1,128 @@
+) : MovieRepository {
+    private val moviesByCategory: Map<MovieCategory, Flow<List<Movie>>> = MovieCategory.values()
+        .associateWith { movieCategory ->
+            flow {
+                val movieResponse = when (movieCategory) {
+                    MovieCategory.POPULAR_ON_TV,
+                    MovieCategory.POPULAR_FOR_RENT,
+                    MovieCategory.POPULAR_IN_THEATRES,
+                    MovieCategory.POPULAR_STREAMING -> {
+                        movieService.fetchPopularMovies()
+                    }
+                    MovieCategory.NOW_PLAYING_MOVIES,
+                    MovieCategory.NOW_PLAYING_TV -> {
+                        movieService.fetchNowPlayingMovies()
+                    }
+                    else -> {
+                        movieService.fetchUpcomingMovies()
+                    }
+                }
+                emit(movieResponse.movies)
+            }.flatMapLatest { apiMovies ->
+                movieDao.favorites()
+                    .map { favoriteMovies ->
+                        apiMovies.map { apiMovie ->
+                            apiMovie.toMovie(isFavorite = favoriteMovies.any { it.id == apiMovie.id })
+                        }
+                    }
+            }.shareIn(
+                scope = CoroutineScope(bgDispatcher),
+                started = SharingStarted.WhileSubscribed(1000L),
+                replay = 1
+            )
+        }
+
+    private val favorites = movieDao.favorites().map {
+        it.map { dbFavoriteMovie ->
+            Movie(
+                id = dbFavoriteMovie.id,
+                imageUrl = dbFavoriteMovie.posterUrl,
+                title = "",
+                overview = "",
+                isFavorite = true
+            )
+        }
+    }.shareIn(
+        scope = CoroutineScope(bgDispatcher),
+        started = SharingStarted.WhileSubscribed(1000L),
+        replay = 1
+    )
+
+    override fun movies(movieCategory: MovieCategory): Flow<List<Movie>> =
+        moviesByCategory[movieCategory]!!
+
+    override fun movieDetails(movieId: Int): Flow<MovieDetails> = flow {
+        emit(movieService.fetchMovieDetails(movieId) to movieService.fetchMovieCredits(movieId))
+    }.flatMapLatest { (apiMovieDetails, apiMovieCredits) ->
+        movieDao.favorites()
+            .map { favoriteMovies ->
+                apiMovieDetails.toMovieDetails(
+                    isFavorite = favoriteMovies.any { it.id == apiMovieDetails.id },
+                    crew = apiMovieCredits.crew.map { it.toCrewman() },
+                    cast = apiMovieCredits.cast.map { it.toCast() }
+                )
+            }
+    }.flowOn(bgDispatcher)
+
+    override fun favoriteMovies(): Flow<List<Movie>> = favorites
+
+    private suspend fun findMovie(movieId: Int): Movie? {
+        return MovieCategory.values().map { movieCategory ->
+            movies(movieCategory).first()
+        }.flatten().find { movie -> movie.id == movieId }
```

Comment:
what if we decide to remove some movie category from `moviesByCategory`? This function should find movie in `moviesByCategory` map, therefore it should iterate over its values:

```kotlin
movoiesByCategory.values.forEach {
    // ...
}
```

Also, think about improving performance of this function. Do we need to flat map  `moviesByCategory` to list first? What if the movie we are searching for is first element in the first accessed slot?